presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
+                    .collect(toImmutableList());
+        }
+
+        private TableCommitContext getTableCommitContext(Page page)
+        {
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                Block operatorExecutionContextBlock = page.getBlock(CONTEXT_CHANNEL);
+                if (!operatorExecutionContextBlock.isNull(position)) {
+                    return tableCommitContextCodec.fromJson(operatorExecutionContextBlock.getSlice(position, 0, operatorExecutionContextBlock.getSliceLength(position)).getBytes());
+                }
+            }
+            throw new IllegalStateException("No operator execution context in input page");
+        }
+
+        private static boolean isLastPageForGroupedLifespan(Page page)
+        {
+            return !page.getBlock(ROW_COUNT_CHANNEL).isNull(0);
+        }
+
+        private static class LifespanStage
+        {
+            private final Lifespan lifespan;
+            private final int stageId;
+
+            private LifespanStage(Lifespan lifespan, int stageId)
+            {
+                this.lifespan = requireNonNull(lifespan, "lifespan is null");
+                this.stageId = stageId;
+            }
+
+            public static LifespanStage fromTableCommitContext(TableCommitContext operatorExecutionContext)
+            {
+                return new LifespanStage(operatorExecutionContext.getLifespan(), operatorExecutionContext.getStageId());
+            }
+
+            public Lifespan getLifespan()
+            {
+                return lifespan;
+            }
+
+            public int getStageId()
+            {
+                return stageId;
+            }
+
+            @Override
+            public boolean equals(Object o)
+            {
+                if (this == o) {
+                    return true;
+                }
+                if (!(o instanceof LifespanStage)) {
+                    return false;
+                }
+                LifespanStage that = (LifespanStage) o;
+                return stageId == that.stageId &&
+                        Objects.equals(lifespan, that.lifespan);
+            }
+
+            @Override
+            public int hashCode()
+            {
+                return Objects.hash(lifespan, stageId);
+            }
+
+            @Override
+            public String toString()
+            {
+                return toStringHelper(this)
+                        .add("lifespan", lifespan)
+                        .add("stageId", stageId)
+                        .toString();
+            }
+        }
+
+        private static class LifespanStageState
+        {
+            private long rowCount;
+            private ImmutableList.Builder<Slice> fragmentBuilder = ImmutableList.builder();
+            private ImmutableList.Builder<Page> statisticsPages = ImmutableList.builder();
+
+            public void update(Page page, boolean storeStatisticPages)
```

Comment:
What's the purpose for `storeStatisticPages` ?

Reply:
We could blindly store all statistic pages as well. I added this because for unrecoverable execution, we actually don't need to store them as they are available as the pages come in given no lifespan commit is needed.

Reply:
After a second thought, I think remove `storeStatisticPages` would make this more readable. :)